### PR TITLE
fix issue 20007 - Add the `__COUNTER__` primary expression

### PR DESCRIPTION
**Pull request by ghost**
This is the second attempt. The [first](https://github.com/dlang/dmd/pull/10093) used a `__trait`. Note that a stuff like this

```d
enum int[3] tup = iota(3).map!(a => __COUNTER__).array;
```

un~~expectedly~~intuitively gives an array of 3 identical values (just like in the first PR) and a stuff like that

```d
template counterT(int v)
{
    enum counterT = __COUNTER__;
}   
enum tup = staticMap!(counterT, aliasSeqOf!(iota(3))); 
```

gives 3 different values but less intuitive to write. But if accepted I plan to add a helper in the library, for example `genUniqueIdentifiers`, so that a demonstration exists.

### CODE REVIEW

**Comment by Basile-z** [comment]
Opened for reviews now. Note that I'm sorry to propose again a language change without DIP but this time it's because this should have been a simple `__trait` initially.

File: changelog/__counter__.dd [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Added the `__COUNTER__` primary expression.
+
+The new `__COUNTER__` special token is used to create a counter expression.
+It is set at compile to a `size_t` constant taken from a compiler side counter that is incremented on each use.
+This counter is the same counter that is used during compilation to generate unique identifiers (to name temporaries for example)
```
Comment (Basile-z) [inline changelog/__counter__.dd]: same one

File: test/runnable/test_counter.d [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+{
+    mixin(`enum uID = "myVar`, __COUNTER__, `";`);
+    mixin("int ", uID, " = 42;");
+    mixin("assert(", uID, " == 42);");
+}
+
+void test5()
+{
+    int i;
+    static assert(__traits(compiles, i = __COUNTER__));
```
Comment (Basile-z) [inline test/runnable/test_counter.d]: `int = size_t` seems to pass on amd64 ? Very strange.

Comment (Basile-z) [reply test/runnable/test_counter.d]: The type may be optimized from the constant value, it's like

```d
    int i;
    enum size_t s = 8;
    static assert(__traits(compiles, i = s));     
```

which compiles too

**Comment by Basile-z** [comment]
The way it's done now makes me think that the same could be achieved in the parser. This would also make the constant value more predictable since the compiler will not generate identifiers until semantic pass.

File: src/dmd/expression.h [hunk 3]
```diff
@@ -1299,6 +1299,18 @@ class DefaultInitExp : public Expression
...
 
+/**
+ * Primary expression for `__COUNTER__`, interfacing the counter used
+ * when generating unique identifiers.
+ */
+class CounterExp : public Expression
+{
+    /// The unique value, assigned during expression semantic.
+    IntegerExp *value;
+    void accept(Visitor v);
```
Comment (ibuclaw) [inline src/dmd/expression.h]: `Visitor *v`

**Review by ibuclaw — DISMISSED** [review]

**Comment by UplinkCoder** [comment]
Looks very good.

**Comment by rainers** [comment]
Please add a test case that shows the behavior with templates.

With the current semantics it will result probably in different counter values if a template is instantiated with the same parameters by two different compilations, i.e. inconsistent symbols or code. I doubt we want to deal with that. Maybe it should just be a counter in the parser?

**Comment by rainers** [comment]
> Almost but unrolling in `static foreach` would be affected so it must be a real expression.

`static foreach` already has a counter builtin, so combining that with `__COUNTER__` should work.

**Comment by Basile-z** [comment]
@thewilsonator : tag as 'Needs works'  + 'Needs Test' please, to prevent accidental merging.

**Comment by thewilsonator** [comment]
Done.

**Pull request closed without merging**